central/policy/datastore/datastore_impl.go

```diff
@@ -235,8 +235,11 @@ func (ds *datastoreImpl) AddPolicy(ctx context.Context, policy *storage.Policy)
 	// Stash away the category names, since they need to be erased on storage. But the policy insert must happen first,
 	// to get an ID, to satisfy foreign key constraints when policy category edges are added.
 	policyCategories := policy.GetCategories()
-	policy.Categories = []string{}
-	err = ds.storage.Upsert(ctx, policy)
+	// Make sure to reset the policy categories field on a clone before upserting; otherwise the given reference
+	// will be changed and information lost when the reference is being kept in-memory (like in policy sets).
+	clonedPolicy := policy.Clone()
+	clonedPolicy.Categories = []string{}
+	err = ds.storage.Upsert(ctx, clonedPolicy)
```

Comment:
The same reference to the underlying policy object is upserted within [the lifecycle manager's policy sets](https://github.com/stackrox/stackrox/blob/master/central/policy/service/service_impl.go#L570-L580); hence the clone here.

Reply:
In this case, the outcome is the same as the expectation (unless there is a developer bug) since we do not anticipate object IDs changing once assigned. However, for consistency and to remove the confusion, I would recommend working off of the cloned object.